File: 30_days_of_code/golang/04_class_vs_instance/main.go
package main
import "fmt"

type person struct {
    age int
}

func (p person) NewPerson(initialAge int) person {
    if initialAge < 0 {
        fmt.Println("Age is not valid, setting age to 0.")
        initialAge = 0
    }

    p.age = initialAge
    return p
}

func (p person) amIOld() (message string) {
    switch age := p.age; {
    case age < 13:
        message = "You are young."
    case age >= 13 && age < 18:
        message = "You are a teenager."
    default:
        message = "You are old."
    }

    return
}

func (p person) yearPasses() person {
    p.age += 1

    return p
}

func main() {
    var T, age int

    fmt.Scan(&T)

    for i := 0; i < T; i++ {
        fmt.Scan(&age)
        p := person{age: age}
        p = p.NewPerson(age)
        p.amIOld()

        for j := 0; j < T; j++ {
            p = p.yearPasses()
        }

        message := p.amIOld()
        fmt.Println(message)
    }
}

File: 30_days_of_code/ruby/04_class_vs_instance/main.rb
class Person
  attr_accessor :age

  def initialize(initial_age)
    if initial_age < 0
      initial_age = 0
      puts 'Age is not valid, setting age to 0.'
    end

    self.age = initial_age
  end

  def am_old?
    message = if age < 13
                'You are young.'
              elsif age >= 13 && age < 18
                'You are a teenager.'
              else
                'You are old.'
              end

    puts message
    message
  end

  def year_passes
    self.age += 1
  end
end
